src/Tools/ExternalAccess/Razor/Features/Cohost/Handlers/SemanticTokensRange.cs

```diff
@@ -13,25 +13,28 @@ namespace Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost.Handlers
 {
     internal static class SemanticTokensRange
     {
+        public static void RegisterRefresh(RazorCohostRequestContext requestContext)
+        {
+            var refreshQueue = requestContext.GetRequiredService<SemanticTokensRefreshQueue>();
+            refreshQueue.AllowRazorRefresh = true;
+        }
+
+        public static Task TryEnqueueRefreshComputationAsync(RazorCohostRequestContext requestContext, Project project, CancellationToken cancellationToken)
```

Comment:
so it looks like you're allowing Razor to manually call a refresh and opting in to the Roslyn automatic refreshing done as part of the queue.   Can you elaborate on why both are required (i.e. why this can't only be done on the Razor side, or only be done on the Roslyn side)?

Reply:
I see.  So this is something that doesn't work for Razor because the handler on our side runs in-proc.  If it ran out of proc instead you could just rely on the request to compute semantic tokens to trigger the refresh as well.